src/mozilla_sec_eia/ex_21/inference.py

```diff
@@ -0,0 +1,366 @@
+        encoding = model_inputs["encoding"]
+        if torch.cuda.is_available():
+            encoding.to("cuda")
+            self.model.to("cuda")
+        # since we're doing inference, we don't need gradient computation
+        with torch.no_grad():
+            output = self.model(**encoding)
+            return {
+                "logits": output.logits,
+                "predictions": output.logits.argmax(-1).squeeze().tolist(),
+                "raw_encoding": model_inputs["raw_encoding"],
+                "example": model_inputs["example"],
+            }
+
+    def postprocess(self, all_outputs):
+        """Return logits, model predictions, and the extracted dataframe."""
+        logits = all_outputs["logits"]
+        predictions = all_outputs["logits"].argmax(-1).squeeze().tolist()
+        output_df = self.extract_table(all_outputs)
+        return logits, predictions, output_df
+
+    def extract_table(self, all_outputs):
+        """Extract a structured table from a set of inference predictions.
+
+        This function essentially works by stacking bounding boxes and predictions
+        into a dataframe and going from left to right and top to bottom. Then, every
+        every time a new subsidiary entity is encountered, it assigns a new group or
+        "row" to that subsidiary. Next, location and ownership percentage words/labeled
+        entities in between these subsidiary groups are assigned to a subsidiary row/group.
+        Finally, this is all formatted into a dataframe with an ID column from the original
+        filename and a basic cleaning function normalizes strings.
+        """
+        predictions = all_outputs["predictions"]
+        encoding = all_outputs["raw_encoding"]
+        example = all_outputs["example"]
+
+        token_boxes_tensor = encoding["bbox"].flatten(start_dim=0, end_dim=1)
+        predictions_tensor = torch.tensor(predictions)
+        mode_predictions = get_flattened_mode_predictions(
+            token_boxes_tensor, predictions_tensor
+        )
+        token_boxes = encoding["bbox"].flatten(start_dim=0, end_dim=1).tolist()
+        predicted_labels = [
+            self.model.config.id2label[pred] for pred in mode_predictions
+        ]
+        simple_preds = [iob_to_label(pred).lower() for pred in predicted_labels]
+
+        df = pd.DataFrame(data=token_boxes, columns=BBOX_COLS)
+        df.loc[:, "iob_pred"] = predicted_labels
+        df.loc[:, "pred"] = simple_preds
+        invalid_mask = (
+            (df["top_left_x"] == 0)
+            & (df["top_left_y"] == 0)
+            & (df["bottom_right_x"] == 0)
+            & (df["bottom_right_y"] == 0)
+        )
+        df = df[~invalid_mask]
+        # we want to get actual words on the dataframe, not just subwords that correspond to tokens
+        # so we merge the original words onto our dataframe on bbox coordinates
+        words_df = pd.DataFrame(data=example["bboxes"], columns=BBOX_COLS)
```

Comment:
Do all subwords from the same complete word have the same bounding box?

Reply:
Yep, because the model isn't actually OCR-ing anything, I'm just passing in bounding boxes that have been created from a PDF parsing package.